lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -356,6 +374,221 @@ void ESIPortsPass::runOnOperation() {
   build = nullptr;
 }
 
+/// Return a attribute with 'rtl.name' appended with suffix.
+static DictionaryAttr appendToRtlName(DictionaryAttr base, StringRef suffix) {
+  auto *ctxt = base.getContext();
+  SmallString<32> retName(suffix);
+  if (auto nameAttr = getRTLNameAttr(base.getValue())) {
+    retName = nameAttr.getValue();
+    retName.append(suffix);
+  }
+  NamedAttribute retNameAttr = NamedAttribute(Identifier::get("rtl.name", ctxt),
+                                              StringAttr::get(ctxt, retName));
+  return DictionaryAttr::get(ctxt, {retNameAttr});
+}
+
+/// Convert all input and output ChannelPorts into valid/ready wires. Try not to
+/// change the order and materialize ops in reasonably intuitive locations.
+bool ESIPortsPass::updateFunc(RTLModuleOp mod) {
+  auto *ctxt = &getContext();
+  auto funcType = mod.getType();
+  // Build ops in the module.
+  ImplicitLocOpBuilder modBuilder(mod.getLoc(), mod.getBody());
+  Type i1 = modBuilder.getI1Type();
+
+  // Get information to be used later on.
+  SmallVector<DictionaryAttr, 8> oldArgAttrs;
+  mod.getAllArgAttrs(oldArgAttrs);
+  SmallVector<DictionaryAttr, 8> oldResultAttrs;
+  mod.getAllResultAttrs(oldResultAttrs);
+  rtl::OutputOp outOp =
+      dyn_cast<rtl::OutputOp>(mod.getBodyBlock()->getTerminator());
+
+  bool updated = false;
+
+  // Reconstruct the list of operand types, changing the type whenever an ESI
+  // port is found. Keep the argument attributes, apply the ESI ports attributes
+  // to the data port only.
+  SmallVector<Type, 16> newArgTypes;
+  SmallVector<DictionaryAttr, 16> newArgAttrs;
+  // 'Ready' signals are outputs. Remember them for later when we deal with the
+  // returns.
+  SmallVector<std::pair<Value, DictionaryAttr>, 8> newReadySignals;
+  for (size_t i = 0, e = funcType.getNumInputs(); i < e; ++i) {
+    Type argTy = funcType.getInput(i);
+
+    auto chanTy = argTy.dyn_cast<ChannelPort>();
+    if (!chanTy) {
+      // If not ESI, pass through.
+      newArgTypes.push_back(argTy);
+      newArgAttrs.push_back(oldArgAttrs[i]);
+      continue;
+    }
+
+    // When we find one, add a data and valid signal to the new args.
+    newArgTypes.push_back(chanTy.getInner());
+    newArgTypes.push_back(i1);
+    newArgAttrs.push_back(oldArgAttrs[i]);
+    newArgAttrs.push_back(appendToRtlName(oldArgAttrs[i], "_valid"));
+    // Add the BlockArguments.
+    Value data = mod.front().insertArgument(i, chanTy.getInner());
+    Value valid = mod.front().insertArgument(i + 1, i1);
+    // Build the ESI wrap operation to translate the lowered signals to what
+    // they were. (A later pass takes care of eliminating the ESI ops.)
+    auto wrap = modBuilder.create<WrapValidReady>(data, valid);
+    // Replace uses of the old ESI port argument with the new one from the wrap.
+    mod.front().getArgument(i + 2).replaceAllUsesWith(wrap.chanOutput());
+    mod.front().eraseArgument(i + 2); // Delete the ESI port block argument.
```

Comment:
Does erasing the arguments in place throw off the iteration? It seems a lot like LowerTypes, which faces an analogous issue. It only removes arguments at the and so the iteration knows where it is in the arguments. A test with more than one channel will probably surface this if it is indeed an issue.

Reply:
I don't think it'll throw off the loop iteration since I'm not iterating through the block arguments -- I'm iterating through the func type inputs. There is a bug there, however. I'm assuming that `i` is both the function input number and the block argument number but after I do this edit, that's no longer true. I'll have to account for that.